### ISSUE TITLE
feature-request: allow package managers to define a version breaking path in package.json

### ISSUE THREAD

**Issue by jsamr** [npm/npm#19231]
#### What's the feature?

Package managers can now decide which part of the version number indicates possible breaking changes, i.e. "incompatible API changes" as per [the semver convention definition](https://semver.org/).

To do so, they can add the following property to `package.json` file<sup>2</sup>:

```json
{
  "breakingPath": "MINOR"
}
```

It's default value is `"MAJOR"` and possible values are:
- `MAJOR`
- `MINOR`
- `PATCH`

#### What problem is the feature intended to solve?

Some libraries and tools purposefully don't follow semver conventions for marketing reasons. A good example of this is Typescript<sup>1</sup>. But this practice upset repo maintainers which need to remember which repo is following semver and which doesn't.

I hope such a feature would reconcile marketing and dev worlds ;-)

#### What yet needs to be discussed

It would seem legit to consider what are the semantics of `^` and `~` with such an addition.
I propose that 

- caret means "allow non-breaking" 
- tilde means "allow patch if exists, otherwise lock this version"
- "2.X" on a module which has `breakingPath` set to `MINOR` would still allow breaking changes

There might be more complexity then I think in such a proposal ; looking forwards for feedback.

-------------------------------

<sup>1</sup> [here is the full argument about Typescript following semver](https://github.com/Microsoft/TypeScript/issues/14116)
<sup>2</sup> the name of this property is only a suggestion, one might be more inspired then I am

**Comment by eddieajau** [comment]
Personally I think to publish to npm you should agree to support SemVer as the standard.

I could perhaps see a use for a flag that a package author can add to tell a consumer they are not following SemVer. Perhaps that flag can automatically adjust whether a `^` or `~` (or neither) is used when installing the package. 

In saying that, however, I think it should be hard for publishers to do this because it promotes confusion for consumers. Consumers shouldn't be the ones being inconvenienced by publishers that don't want to follow the npm standard. It's much, much better on everyone if the entire ecosystem just uses SemVer.

**Comment by jsamr** [comment]
@eddieajau Yeah I totally agree about the implicit contract `npm` stands for. But at some point we need to be realistic and provide the best experience for library consumers.
